controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,260 @@
+		Group:    "constraints.gatekeeper.sh",
+		Version:  "v1beta1",
+		Resource: strings.ToLower(constraintKind),
+	}
+
+	constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, constraintName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	constraintMatchKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	contraintSyncOnlyEntries, err := r.getSyncOnlys(constraintMatchKinds)
+	if err != nil {
+		log.Error(err, "Error to get matching kind and apigroup")
+
+		return reconcile.Result{}, nil
+	}
+
+	for _, ce := range *contraintSyncOnlyEntries {
+		table[ce] = true
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+	for key := range table {
+		syncOnlys = append(syncOnlys, key)
+	}
+
+	config.Spec.Sync.SyncOnly = syncOnlys
+	err = r.Update(ctx, config, &client.UpdateOptions{})
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ConstraintStatusReconciler) getSyncOnlys(constraintMatchKinds []interface{}) (
+	*[]v1alpha1.SyncOnlyEntry, error,
+) {
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for _, kind := range constraintMatchKinds {
+		newKind := kind.(map[string]interface{})
+		apiGroups := newKind["apiGroups"].([]interface{})
+		kindsInKinds := newKind["kinds"].([]interface{})
+
+		for _, apiGroup := range apiGroups {
+			for _, kindkind := range kindsInKinds {
+				version, err := r.getAPIVersion(kindkind.(string), apiGroup.(string))
+				if err != nil {
+					return nil, err
+				}
+
+				syncOnlys = append(syncOnlys, v1alpha1.SyncOnlyEntry{
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kindkind.(string),
+				})
+			}
+		}
+	}
+
+	return &syncOnlys, nil
+}
+
+func (r *ConstraintStatusReconciler) getAPIVersion(kind string, apiGroup string) (string, error) {
+	// cool time(10 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		r.discoveryLastRefreshed = time.Now()
+
+		if err != nil {
+			return "", err
+		}
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot Parse Group and version in getApiVersion")
+
+			return "", errors.New("Error in parse Group and version in getApiVersion function")
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	// Get new discoveryInfo, when any resource is not found
+	err := r.refreshDiscoveryInfo()
+	r.discoveryLastRefreshed = time.Now()
+
+	if err != nil {
+		return "", err
+	}
+
+	return "", errors.New("Getting discovery has error")
```

Comment:
An error that can be matched with `errors.Is` may be useful here to determine if the error was because the discovery information didn't have the answer compared to a potentially recoverable error.